zkvm: introduce anchors (#193)
interstellar/slingshot: zkvm/docs/zkvm-spec.md

[`mintime`](#mintime) and [`maxtime`](#maxtime).

### Contract ID

_Contract ID_ is a 32-byte unique identifier of the [contract](#contract-type) and a commitment to its contents (see [Output](#output-structure)):

```
T = Transcript("ZkVM.contractid")
T.commit("contract", output_structure)
id = T.challenge_bytes("id")
```

> Contract ID makes signatures safe against reused [predicate](#predicate) keys:

Reviewer: What does the phrase "predicate key" mean here?

Author: Changed to 
```
Contract ID makes signatures safe against reused [verification keys](#verification-key) in [predicates](#predicate):
```

> id = T.challenge_bytes()

Reviewer: Does `challenge_bytes` not require a string argument like `"id"`?

If not, why is it used in some places?

Author: it requires, i fixed it and not pushed it yet

a signature covers the unique contract ID, therefore preventing its replay against another contract,
even if containing the same [payload](#contract-payload) and predicate.

Uniqueness is provided via the [anchor](#anchor).

### Anchor

_Anchor_ is a 32-byte unique string that provides uniqueness to the [contract ID](#contract-id).

> Anchors can be created by the [`nonce`](#nonce) instruction or generated from previously used unique contract IDs, tracked by the VM via [last anchor](#vm-state):

Reviewer: ```suggestion
Anchors can be created by the [`nonce`](#nonce) instruction or generated from unique contract IDs used earlier in the same transaction. These are tracked by the VM via [last anchor](#vm-state):
```

_Anchor_ is a 32-byte unique string that provides uniqueness to the [contract ID](#contract-id).

Anchors can be created by the [`nonce`](#nonce) instruction or generated from previously used unique contract IDs, tracked by the VM via [last anchor](#vm-state):

1. Nonce contracts have their anchor computed from the nonce parameters (see [`nonce`](#nonce) instruction).

> 2. Claimed UTXOs ([`input`](#input)) carry over the previously created anchor and set the "last anchor" to their [contract ID](#contract-id).

Author: will clarify, thanks

> 3. Transient contract ([`contract`](#contract)) consumes the VM’s [last anchor](#vm-state) and replaces it with its [contract ID](#contract-id).

Reviewer: Is it safe for `contract` to use last-anchor as-is? If so, can you say why? Or should it be ratcheted as with `output`?

Author: WDYT?

> Transient contract sets its ID as an anchor because this can be done only once: the contract is soon destroyed by `signtx`, `call` or `delegate`.
> This is not the case with `output`: if the output’s ID was set as the anchor, it'd be available in two transactions: in this one, for the rest of the execution, and in the future one, when the output is claimed by `input` which sets the same output ID as an anchor.

Reviewer: OK ✔️

4. Newly created outputs ([`output`](#output)) consume the VM’s [last anchor](#vm-state) and replace it with a ratcheted one (because the contract ID will be used as an anchor in a later transaction, via [`input`](#input)).

VM keeps track of the last used anchor and fails if:

1. by the end of the execution, no anchor was used (which means that [transaction ID](#transaction-id) is not unique), or
2. an [`output`](#output) or [`contract`](#output) are invoked before the anchor is set (since they can’t be made unique).

> Note: using a chain of anchors enables the signer to have a predictable anchor computation, when the signable contract is created transiently from a just-opened UTXO, without the need to know about the entire transaction.

Author: How about this?

> Chaining the anchors this way gives flexibility to the signer: if contract A creates contract B, the B's ID can be computed solely from the contents of A, without the knowledge of the entire transaction.

> Note: using a chain of anchors enables the signer to have a predictable anchor computation, when the signable contract is created transiently from a just-opened UTXO, without the need to know about the entire transaction.

Reviewer: Looks good, thanks!

but in the [output](#output-structure) only the [portable types](#portable-types) are allowed.

### Output structure

> Output represents a _snapshot_ of a [contract](#contract-type)

Reviewer: Putting snapshot in italics sort of implies it's a technical term whose definition is nearby. Perhaps it should not be in italics and you can get away with saying something like ..."a snapshot of a contract's state at the moment it executed an `output` instruction."

Author: yeah, makes sense

Author: changed to

```
Output is a serialized [contract](#contract-type):
```

but in the [output](#output-structure) only the [portable types](#portable-types) are allowed.

### Output structure

Output represents a _snapshot_ of a [contract](#contract-type)
and can only contain [portable types](#portable-types).

```
      Output  =  Anchor || Predicate  ||  LE32(k)  ||  Item[0]  || ... ||  Item[k-1]
      Anchor  =  <32 bytes>
   Predicate  =  <32 bytes>
        Item  =  enum { Data, Value }
        Data  =  0x00  ||  LE32(len)  ||  <bytes>
       Value  =  0x01  ||  <32 bytes> ||  <32 bytes>
```

### UTXO

UTXO stands for Unspent Transaction [Output](#output-structure).

> UTXO is uniquely identified by the [contract ID](#contract-id).

Reviewer: ```suggestion
UTXO is uniquely identified by the [ID](#contract-id) of the contract in the output.
```

    * `tx_signature`
    * `cs_proof`
2. Extension flag (boolean)

> 3. Last anchor (optional [anchor](#anchor): `None` or `Some(anchor)`)

Reviewer: Avoid Rust idioms in the spec. I suggest something like:

```suggestion
3. Last [anchor](#anchor) or ∅ if unset.
```

Author: good point

1. [Transaction](#transaction) as provided by the user.
2. Extension flag set to `true` or `false` according to the [transaction versioning](#versioning) rules for the transaction version.

> 3. Last anchor is set to `None`.

Reviewer: ```suggestion
3. Last anchor is set to ∅.
```

(To match some earlier text.)

If the execution finishes successfully, VM performs the finishing tasks:
1. Checks if the stack is empty; fails otherwise.

> 2. Checks if the last anchor is present; fails otherwise.

Reviewer: ```suggestion
2. Checks if the last anchor is set to other than ∅; fails otherwise.
```

Author: this reads strangely to me. Maybe simply  "checks if the last anchor is set"?

Author: ```
2. Checks if the last anchor is set; fails otherwise.
```

    T.commit("metadata", metadata)
    flavor = T.challenge_scalar("flavor")
    ```
6. Checks that the `flv` has unblinded commitment to `flavor`
   by [deferring the point operation](#deferred-point-operations):
    ```
    flv == flavor·B
    ```
7. Adds a 64-bit range proof for the `qty` to the [constraint system](#constraint-system)
   (see [Cloak protocol](../../spacesuit/spec.md) for the range proof definition).
8. Adds an [issue entry](#issue-entry) to the [transaction log](#transaction-log).
9. Creates a [contract](#contract-type) with the value as the only [payload](#contract-payload),
   protected by the predicate `pred`, consuming [VM’s last anchor](#vm-state)
   and replacing it with this contract’s [ID](#contract-id).

The value is now issued into the contract that must be unlocked
using one of the contract instructions: [`signtx`](#signtx), [`delegate`](#delegate) or [`call`](#call).

Fails if:
* `pred` is not a valid [point](#point),
* `flv` or `qty` are not [variable types](#variable-type),

> * VM’s [last anchor](#vm-state) is missing.

Reviewer: ```suggestion
* VM’s [last anchor](#vm-state) is not set.
```

#### input

_prevoutput_ **input** → _contract_

1. Pops a [data](#data-type) `prevoutput` representing the [unspent output structure](#output-structure) from the stack.
2. Constructs a [contract](#contract-type) based on the `prevoutput` data and pushes it to the stack.
3. For each decoded [value](#value-type), quantity variable is allocated first, flavor second.
4. Adds [input entry](#input-entry) to the [transaction log](#transaction-log).
5. Sets the [VM’s last anchor](#vm-state) to the [contract ID](#contract-id) of the claimed 
[UTXO](#utxo).

Fails if the `prevoutput` is not a [data type](#data-type) with exact encoding of an [output structure](#output-structure).

#### output

_items... predicate_ **output:_k_** → ø

1. Pops [`predicate`](#predicate) from the stack.
2. Pops `k` items from the stack.
3. Use the [VM’s last anchor](#vm-state) as an anchor for the [output structure](#output-structure).
4. Update the [VM’s last anchor](#vm-state) with a ratcheted [contract ID](#contract-id):
    ```
    T = Transcript("ZkVM.ratchet-anchor")

>     T.commit("output", output_id)

Reviewer 2: Here, what is `output_id`? Is this the VM's last anchor?

Author: rewrote this part here: 
https://github.com/interstellar/slingshot/pull/193/commits/6b995d569476e1bf2b41f03fd3222482f80403bf

Fails if VM’s [last anchor](#vm-state) is missing.

#### nonce

_pred blockid_ **nonce** → _contract_

1. Pops [32-byte string](#data-type) `blockid` from the stack.
2. Pops [predicate](#predicate) `pred` from the stack.
3. Computes the nonce [anchor](#anchor) value:
    ```
    T = Transcript("ZkVM.nonce")
    T.commit("blockid", blockid)
    T.commit("predicate", pred)
    T.commit("maxtime", LE64(tx.maxtime))
    nonce_anchor = T.challenge_bytes()
    ```
4. Pushes a new [contract](#contract-type) with:
    * an empty [payload](#contract-payload)
    * predicate set to `pred`
    * anchor set to `nonce_anchor`.
5. Sets the [VM’s last anchor](#vm-state) to [contract ID](#contract-id) of that contract.
6. Adds [nonce entry](#nonce-entry) to the [transaction log](#transaction-log) with the `blockid`, transaction [maxtime](#time-bounds) and `nonce_anchor`.

Fails if:
* `blockid` is a not a 32-byte [data string](#data-type).
* `pred` is not a valid [point](#point).

**Discussion**

A nonce serves two purposes:

1. It provides a unique anchor for an issuing transaction if other anchors are not available (i.e., from other values already on the stack);

> 2. It binds the entire transaction to a particular blockchain, protecting not only against cross-blockchain replays, but also potential blockchain forks due to compromise of the old block-signing keys. This is a necessary feature for stateless signing devices that rely on blockchain proofs.

Reviewer: > block-signing keys

Not clear to me that we're going to have those. Perhaps only in the consensus layer, not in the blockchain data model per se.

Author: yeah, that paragraph is very much out of scope. How about: 
```suggestion
2. It binds the entire transaction to a particular blockchain, protecting not only against cross-blockchain replays, but also potential blockchain forks. This is a necessary feature for stateless signing devices that rely on blockchain proofs.
```

Reviewer: That looks good, but note that block signatures are also mentioned in one or two other places.

Author: oops

Blockchain state machine performs the following checks:

> 1. The `blockid` is checked against the IDs of “recent” blocks when the transaction is applied to the blockchain. It must match a recent block, or the initial block of the blockchain.

Reviewer: TxVM also allows it to be 0x00 × 32 bytes. I'm fine with excluding that possibility but first let's make sure we remember why TxVM allows it in the first place. (I don't, do you?)

Author: It was allowed to support hypothetical case where we can do fancy replay of the same transaction on two TxVM chains to do an unspecified shorter cross-chain swap. However, having this possibility, precludes a guarantee that all anchors are globally unique. So if someone somewhere used an zero-bound nonce, its non-uniqueness can shoot someone's foot down the road through the chain of anchor derivation.

Since we are going to make "cross-chain swaps" AKA "payment channels" by the book and be happy with that, I don't think we need to keep this dangerous behavior.

Reviewer: OK

#### input

_prevoutput_ **input** → _contract_

1. Pops a [data](#data-type) `prevoutput` representing the [unspent output structure](#output-structure) from the stack.
2. Constructs a [contract](#contract-type) based on the `prevoutput` data and pushes it to the stack.
3. Adds [input entry](#input-entry) to the [transaction log](#transaction-log).
4. Sets the [VM’s last anchor](#vm-state) to the ratcheted [contract ID](#contract-id):
    ```
    T = Transcript("ZkVM.ratchet-anchor")
    T.commit("old", contract_id)
    new_anchor = T.challenge_bytes("new")
    ```

> Fails if the `prevoutput` is not a [data type](#data-type) with exact encoding of an [output structure](#output-structure).

Reviewer: I think this whole document needs to replace the way it uses "data type." Brought this up [in Slack](https://i10r.slack.com/archives/CCVPR72CR/p1549411686017000?thread_ts=1549408834.013200&cid=CCVPR72CR) originally:

> the way the spec uses the phrase "data type" is causing some cognitive dissonance. e.g. "scalars are encoded as 32-byte data types." those are _strings_ (or _blobs_ or _bytearrays_ or something). a "data type" should be a type, not data

Author: Let's revisit this in another PR. I agree that "data" is confusing, but "blob" and "bytearray" sound clunky and "string" sounds too much like "human text" (e.g. most modern languages have String type for UTF8 text). Motivation for "data" was as opposite to "non-data" linear types like "Values" and "Contracts". Those are not _data types_ because those are _things_. But i don't deny there's confusion alike to would-be one with String-as-bytes VS String-as-humantext.

Reviewer: Sounds good. To get a little pedantic, in the phrase "data type," "type" works as a noun and "data" works as an adjective modifying it; so it's a kind of type, not a kind of data. (The phrase "type data" would be a kind of data, not a kind of type.) My objection is not to the word "data" per se, but to "type" used as a noun to describe data.

Author: Filed: 
https://github.com/interstellar/slingshot/issues

#### output

_items... predicate_ **output:_k_** → ø

1. Pops [`predicate`](#predicate) from the stack.
2. Pops `k` items from the stack.
3. Creates a contract with the `k` items as a payload, the predicate `pred`, and anchor set to the [VM’s last anchor](#vm-state).
4. Adds an [output entry](#output-entry) to the [transaction log](#transaction-log).

> 5. Update the [VM’s last anchor](#vm-state) with the [contract ID](#contract-id) of the new contract.

Reviewer: ```suggestion
5. Updates the [VM’s last anchor](#vm-state) with the [contract ID](#contract-id) of the new contract.
```

Immediate data `k` is encoded as [LE32](#le32).

> Fails if VM’s [last anchor](#vm-state) is missing.

Reviewer: ```suggestion
Fails if VM’s [last anchor](#vm-state) is not set.
```

* `issue` does not consume zero value
* `finalize` does not consume zero value
* `claim/borrow` can produce an arbitrary value and its negative at any point
* Each UTXO ID is defined as `Hash(contract)`, that is contents of the contract are unique due to a contract’s anchor.
* Transaction ID is a hash of the finalized log.
* Newly created contracts consume the "last anchor" tracked by VM and set it to their contract ID.
* When VM finishes, it checks that "last anchor" is present, which means that the log contains either an [input](#input-entry) or a [nonce](#nonce-entry).
* Inputs and outputs are represented in the log as contract IDs which the blockchain state machine should remove/insert accordingly.

**Pros:**

The values are simply pedersen commitments (Q,A) (quantity, flavor), without any extra payload.
This makes handling them much simpler: we can issue, finalize and even “claim” a value in a straightforward manner.

**Con:**

> UTXO IDs are not generally known without fuller view on transaction flow. This could be not a big deal, as we cannot really plan for the next transaction until this one is fully formed and published. Also, in a joint proof scenario, it’s even less reliable to plan the next payment until the MPC is completed, so requirement to wait till transaction ID is determined may not be a big deal.

Reviewer: Expand MPC?

Author: ```suggestion
UTXO IDs are not generally known without fuller view on transaction flow. This could be not a big deal, as we cannot really plan for the next transaction until this one is fully formed and published. Also, in a joint proof scenario, it’s even less reliable to plan the next payment until the multi-party protocol is completed, therefore waiting for transaction ID to be determined becomes a requirement.
```

Reviewer: :+1:
